### PR TITLE
Adds packet traffic class hierarchy and scheduler.

### PR DESCRIPTION
**Pull request by barath**

### CODE REVIEW

**Comment by azanwar** [comment]
test comment

**Comment by hao-liu-1117** [comment]
test comment

**Comment by floan** [comment]
Test comment

**Comment by williamsuppiger** [comment]
test

**Comment by gsusanto** [comment]
test comment

**Comment by eye942** [comment]
test

**Comment by gy-aiden** [comment]
test

**Comment by JanzenLiu** [comment]
test

File: scheduler.h [hunk 1]
```diff
@@ -0,0 +1,399 @@
...
+
+namespace bess {
+
+struct sched_stats {
+  resource_arr_t usage;
+  uint64_t cnt_idle;
+  uint64_t cycles_idle;
+};
+
+class Scheduler;
```
Comment (floan) [inline scheduler.h:52]: Forward Declaration - Not good coding style

File: scheduler.h [hunk 2]
```diff
@@ -0,0 +1,399 @@
...
+    }
+  };
+
+  SchedWakeupQueue() : q_() {}
+
+  // Adds the given traffic class to those that are considered blocked.
+  void Add(TrafficClass *c) { q_.push(c); }
+
+  // Removes the given traffic class from the blocked list.
+  void Remove(const TrafficClass *c) {
```
Comment (Edward-Chow) [inline scheduler.h:70]: Is that put the implementation of the member function in .cc file a better choice?

File: traffic_class.h [hunk 3]
```diff
@@ -0,0 +1,783 @@
...
+}  // namespace traffic_class_initializer_types
+
+using namespace traffic_class_initializer_types;
+
+const std::string TrafficPolicyName[NUM_POLICIES] = {
+    "priority", "weighted_fair", "round_robin", "rate_limit", "leaf"};
+
+const std::unordered_map<std::string, enum resource_t> ResourceMap = {
+    {"count", RESOURCE_COUNT},
+    {"cycle", RESOURCE_CYCLE},
```
Comment (azanwar) [inline traffic_class.h:131]: convert 4 spaces to 2

File: scheduler.h [hunk 4]
```diff
@@ -0,0 +1,399 @@
...
+
+  // Runs the scheduler loop forever.
+  void ScheduleLoop() override {
+    uint64_t now;
+    // How many rounds to go before we do accounting.
+    const uint64_t accounting_mask = 0xff;
+    static_assert(((accounting_mask + 1) & accounting_mask) == 0,
+                  "Accounting mask must be (2^n)-1");
+
+    this->checkpoint_ = now = rdtsc();
```
Comment (floan) [inline scheduler.h:234]: Chaining operator= ??

File: scheduler.h [hunk 5]
```diff
@@ -0,0 +1,399 @@
...
+      ctx->current_tsc = this->checkpoint_;  // Tasks see updated tsc.
+      ctx->current_ns = this->checkpoint_ * this->ns_per_cycle_;
+      current_worker.set_current_tsc(ctx->current_tsc);
+      current_worker.set_current_ns(ctx->current_ns);
+
+      ctx->task = leaf->task();
+      ctx->silent_drops = 0;
+
+      // Run.
+      auto ret = (*ctx->task)(ctx);
```
Comment (floan) [inline scheduler.h:272]: What is happening here?

File: traffic_class.cc [hunk 6]
```diff
@@ -0,0 +1,541 @@
...
+size_t TrafficClass::Size() const {
+  size_t ret = 1;  // itself
+  for (const auto *child : Children()) {
+    ret += child->Size();
+  };
+  return ret;
+}
+
+int TrafficClass::WorkerId() const {
+  for (int wid = 0; wid < Worker::kMaxWorkers; wid++) {
```
Comment (gy-aiden) [inline traffic_class.cc:53]: always use prefix increment, ++wid

File: scheduler.h [hunk 7]
```diff
@@ -0,0 +1,399 @@
...
+// needed for scheduling.
+class Scheduler {
+ public:
+  explicit Scheduler(TrafficClass *root = nullptr)
+      : root_(root),
+        default_rr_class_(),
+        wakeup_queue_(),
+        stats_(),
+        checkpoint_(),
+        ns_per_cycle_(1e9 / tsc_hz) {}
```
Comment (hao-liu-1117) [inline scheduler.h:92]: '1e9' could be a constant variable with name.

File: traffic_class.h [hunk 8]
```diff
@@ -0,0 +1,783 @@
...
+
+// The amount of a resource allocated to a class.
+typedef int32_t resource_share_t;
+
+struct tc_stats {
+  resource_arr_t usage;
+  uint64_t cnt_throttled;
+};
+
+class Scheduler;
```
Comment (floan) [inline traffic_class.h:86]: Redefining a class? Why not just include scheduler.h?

File: traffic_class.h [hunk 9]
```diff
@@ -0,0 +1,783 @@
...
+#include "utils/extended_priority_queue.h"
+#include "utils/simd.h"
+#include "utils/time.h"
+
+using bess::utils::extended_priority_queue;
+
+namespace bess {
+
+// A large default priority.
+#define DEFAULT_PRIORITY 0xFFFFFFFFu
```
Comment (Edward-Chow) [inline traffic_class.h:55]: Can we use constexpr instead of marco here to define constant value?

File: traffic_class.h [hunk 10]
```diff
@@ -0,0 +1,783 @@
...
+
+/* acc += x */
+#define ACCUMULATE(acc, x)                                \
+  {                                                       \
+    uint64_t *p1 = acc;                                   \
+    uint64_t *p2 = x;                                     \
+    for (int index = 0; index < NUM_RESOURCES; ++index) { \
+      p1[index] += p2[index];                             \
+    }                                                     \
+  }
```
Comment (floan) [inline traffic_class.h:149]: Don't understand what this is, why not redefine operator+= for this class?

File: traffic_class.h [hunk 11]
```diff
@@ -0,0 +1,783 @@
...
+// An array of counters for all resource types.
+typedef uint64_t resource_arr_t[NUM_RESOURCES];
+
+// The priority of a traffic class.
+typedef uint32_t priority_t;
+
+// The amount of a resource allocated to a class.
+typedef int32_t resource_share_t;
+
+struct tc_stats {
```
Comment (Edward-Chow) [inline traffic_class.h:81]: name of struct should be camelcase

File: traffic_class.h [hunk 12]
```diff
@@ -0,0 +1,783 @@
...
+enum resource_t {
+  RESOURCE_COUNT = 0,  // Count of how many times scheduled
+  RESOURCE_CYCLE,      // CPU cycles
+  RESOURCE_PACKET,     // Packets set
+  RESOURCE_BIT,        // Bits sent
+  NUM_RESOURCES,       // Sentinel. Also used to indicate "no resource".
+};
+
+// An array of counters for all resource types.
+typedef uint64_t resource_arr_t[NUM_RESOURCES];
```
Comment (Edward-Chow) [inline traffic_class.h:73]: name of typedef should be camelcase

File: traffic_class.h [hunk 13]
```diff
@@ -0,0 +1,783 @@
...
+      : parent_type_(NUM_POLICIES), child_(child) {}
+  TrafficClass *child() { return child_; }
+  TrafficPolicy parent_type() { return parent_type_; }
+
+ protected:
+  TCChildArgs(TrafficPolicy parent_type, TrafficClass *child)
+      : parent_type_(parent_type), child_(child) {}
+
+ private:
+  const TrafficPolicy parent_type_;
```
Comment (Edward-Chow) [inline traffic_class.h:163]: const variable should be named starting with 'k'

File: traffic_class_bench.cc [hunk 14]
```diff
@@ -0,0 +1,237 @@
...
+
+#include "scheduler.h"
+#include "traffic_class.h"
+
+#define CT TrafficClassBuilder::CreateTree
+
+using namespace bess;
+using namespace bess::traffic_class_initializer_types;
+
+namespace {
```
Comment (floan) [inline traffic_class_bench.cc:47]: No name after the namespace?

File: traffic_class_bench.cc [hunk 15]
```diff
@@ -0,0 +1,237 @@
...
+    dummy_ = new DummyModule;
+
+    TrafficClass *root =
+        CT("root", {PRIORITY},
+           {{0, CT("weighted", {WEIGHTED_FAIR, resource}, {})}});
+    s_ = new DefaultScheduler(root);
+    WeightedFairTrafficClass *weighted =
+        static_cast<WeightedFairTrafficClass *>(
+            TrafficClassBuilder::Find("weighted"));
+    for (int i = 0; i < num_classes; i++) {
```
Comment (floan) [inline traffic_class_bench.cc:80]: Small nit but switch to ++i for consistency.

**Review by floan — COMMENTED** [review]
General question: Why are you using namespaces everywhere? Especially if you are defining classes (and subsequently methods) in those namespaces. Couldn't you just uniquely identify the methods and classes by their names?

**Comment by williamsuppiger** [comment]
- traffic_class.h:142 macro
- TrafficClass is a node in Tree
- traffic_class_test.cc:61 unit test
- traffic_class_test.cc:42 stay away from Macros
- doesn’t test edge cases (only expected cases)
- traffic_class_test.cc compiler directive? What does it do?
- scheduler.h:94 TODO

File: traffic_class.h [hunk 16]
```diff
@@ -0,0 +1,783 @@
...
+
+using bess::utils::extended_priority_queue;
+
+namespace bess {
+
+// A large default priority.
+#define DEFAULT_PRIORITY 0xFFFFFFFFu
+
+// Share is defined relatively, so 1024 should be large enough
+#define STRIDE1 (1 << 20)
```
Comment (JanzenLiu) [inline traffic_class.h:58]: The comment mentioned 1024 but I think (1 << 20) is not 1024?

File: traffic_class.h [hunk 17]
```diff
@@ -0,0 +1,783 @@
...
+// The amount of a resource allocated to a class.
+typedef int32_t resource_share_t;
+
+struct tc_stats {
+  resource_arr_t usage;
+  uint64_t cnt_throttled;
+};
+
+class Scheduler;
+class SchedWakeupQueue;
```
Comment (JanzenLiu) [inline traffic_class.h:87]: Does `Sched` stands for `Scheduler`? I think maybe using the full word will be better.

File: traffic_class.h [hunk 18]
```diff
@@ -0,0 +1,783 @@
...
+enum RateLimitFakeType {
+  RATE_LIMIT = 0,
+};
+enum LeafFakeType {
+  LEAF = 0,
+};
+
+}  // namespace traffic_class_initializer_types
+
+using namespace traffic_class_initializer_types;
```
Comment (JanzenLiu) [inline traffic_class.h:124]: I think using namespace in header files is not recommended.

File: traffic_class.h [hunk 19]
```diff
@@ -0,0 +1,783 @@
...
+/* acc += x */
+#define ACCUMULATE(acc, x)                                \
+  {                                                       \
+    uint64_t *p1 = acc;                                   \
+    uint64_t *p2 = x;                                     \
+    for (int index = 0; index < NUM_RESOURCES; ++index) { \
+      p1[index] += p2[index];                             \
+    }                                                     \
+  }
+
```
Comment (JanzenLiu) [inline traffic_class.h:150]: I think macro function is discouraged?

File: traffic_class.h [hunk 20]
```diff
@@ -0,0 +1,783 @@
...
+  // this TC.
+  size_t Size() const;
+
+  virtual std::vector<TrafficClass *> Children() const = 0;
+
+  // Returns the root of the tree this class belongs to.
+  // Expensive in that it is recursive, so do not call from
+  // performance-sensitive code.
+  const TrafficClass *Root() const { return parent_ ? parent_->Root() : this; }
+  TrafficClass *Root() { return parent_ ? parent_->Root() : this; }
```
Comment (JanzenLiu) [inline traffic_class.h:183]: Is there any reason to have two versions (one const and one non-const) of Root()?

File: traffic_class.h [hunk 21]
```diff
@@ -0,0 +1,783 @@
...
+  bool blocked() const { return blocked_; }
+
+  TrafficPolicy policy() const { return policy_; }
+
+ protected:
+  friend PriorityTrafficClass;
+  friend WeightedFairTrafficClass;
+  friend RoundRobinTrafficClass;
+  friend RateLimitTrafficClass;
+  friend class LeafTrafficClass;
```
Comment (JanzenLiu) [inline traffic_class.h:218]: friend classes are also declared in the private section, is there any difference where (public/protected/private) to declare them?

File: traffic_class.h [hunk 22]
```diff
@@ -0,0 +1,783 @@
...
+
+  explicit PriorityTrafficClass(const std::string &name)
+      : TrafficClass(name, POLICY_PRIORITY), first_runnable_(0), children_() {}
+
+  ~PriorityTrafficClass();
+
+  std::vector<TrafficClass *> Children() const override;
+
+  // Returns true if child was added successfully.
+  bool AddChild(TrafficClass *child, priority_t priority);
```
Comment (JanzenLiu) [inline traffic_class.h:308]: AddChild() is not declared in TrafficClass class, but every derived class has a AddChild() method. I am aware that the AddChild() method of each derived class may have different arguments, but is there any better way to do that? (so that we have a unified interface defined in the base class)
